apollo-router-core/src/layers/apq.rs

```diff
@@ -0,0 +1,427 @@
+use crate::{plugin_utils::structures::RouterResponseBuilder, RouterRequest, RouterResponse};
+use futures::future::BoxFuture;
+use moka::sync::Cache;
+use serde::Deserialize;
+use serde_json_bytes::json;
+use sha2::{Digest, Sha256};
+use std::task::Poll;
+use tower::{BoxError, Layer, Service};
+
+#[derive(Deserialize, Clone, Debug)]
+pub struct PersistedQuery {
+    pub version: u8,
+    #[serde(rename = "sha256Hash")]
+    pub sha256hash: String,
+}
+
+#[derive(Clone)]
+pub struct APQ {
+    cache: Cache<Vec<u8>, String>,
+    response_builder: RouterResponseBuilder,
+}
+
+impl APQ {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            cache: Cache::new(capacity),
+            response_builder: RouterResponseBuilder::new().push_error(crate::Error {
+                message: "PersistedQueryNotFound".to_string(),
+                locations: Default::default(),
+                path: Default::default(),
+                extensions: serde_json_bytes::from_value(json!({
+                      "code": "PERSISTED_QUERY_NOT_FOUND",
+                      "exception": {
+                      "stacktrace": [
+                          "PersistedQueryNotFoundError: PersistedQueryNotFound",
+                      ],
+                  },
+                }))
+                .unwrap(),
+            }),
+        }
+    }
+}
+pub struct APQService<S>
+where
+    S: Service<RouterRequest>,
+{
+    service: S,
+    apq: APQ,
+}
+
+impl<S> APQService<S>
+where
+    S: Service<RouterRequest>,
+{
+    pub fn new(service: S, capacity: usize) -> Self {
+        Self {
+            service,
+            apq: APQ::with_capacity(capacity),
+        }
+    }
+}
+
+impl<S> Layer<S> for APQ
+where
+    S: Service<RouterRequest, Response = RouterResponse>,
+{
+    type Service = APQService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        APQService {
+            apq: self.clone(),
+            service,
+        }
+    }
+}
+
+impl<S> Service<RouterRequest> for APQService<S>
+where
+    S: Service<RouterRequest, Response = RouterResponse, Error = BoxError>,
+    S::Future: Send + 'static,
+{
+    type Response = <S as Service<RouterRequest>>::Response;
+
+    type Error = <S as Service<RouterRequest>>::Error;
+
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: RouterRequest) -> Self::Future {
+        let apq = self.apq.clone();
```

Comment:
Doesn't this mean we'll get a new APQ struct which we'll then throw away at end of scope?

I added a comment above (line 49) about providing some form of interior mutability.

(We may not need interior mutability, that's just one approach and depends on our locking requirements, but we need something which means we aren't cloning.)

This diff, for instance, avoids cloning `apq` and relies on cloning the ResponseBuilder if required. I'm not sure if the cache will work properly without synchronisation though:

```
diff --git a/apollo-router-core/src/layers/apq.rs b/apollo-router-core/src/layers/apq.rs
index c833d6f..f130c28 100644
--- a/apollo-router-core/src/layers/apq.rs
+++ b/apollo-router-core/src/layers/apq.rs
@@ -91,7 +91,7 @@ where
     }
 
     fn call(&mut self, mut req: RouterRequest) -> Self::Future {
-        let apq = self.apq.clone();
+        // let apq = self.apq;
 
         let req = {
             let maybe_query_hash: Option<Vec<u8>> = req
@@ -112,19 +112,21 @@ where
                 (Some(query_hash), Some(query)) => {
                     if query_matches_hash(query.as_str(), query_hash.as_slice()) {
                         tracing::trace!("apq: cache insert");
-                        apq.cache.insert(query_hash, query);
+                        self.apq.cache.insert(query_hash, query);
                     } else {
                         tracing::warn!("apq: graphql request doesn't match provided sha256Hash");
                     }
                 }
                 (Some(apq_hash), _) => {
-                    if let Some(cached_query) = apq.cache.get(&apq_hash) {
+                    if let Some(cached_query) = self.apq.cache.get(&apq_hash) {
                         tracing::trace!("apq: cache hit");
                         req.http_request.body_mut().query = Some(cached_query);
                     } else {
                         tracing::trace!("apq: cache miss");
-                        let res = apq
+                        let res = self
+                            .apq
                             .response_builder
+                            .clone()
                             .with_context(req.context.with_request(req.http_request.into()))
                             .build();
                         return Box::pin(async move { Ok(res) });
```

Reply:
it's not super explicit, it should be once i express the send / sync requirement more clearly, but there's an arc under the hood in the moka implementation, so we're effectively hitting the same cache here